File: typescript/packages/core/src/core/discord/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { DiscordAccessor } from '../../accessor/discord.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import { PathSpec } from '../../types.ts'
import { downloadFile } from './files.ts'
import { getHistoryJsonl } from './history.ts'
import { listMembers } from './members.ts'
import { readdir as discordReaddir } from './readdir.ts'
import { stripSlash } from '../../util/slash.ts'

const encoder = new TextEncoder()

function fileNotFound(key: string): Error {
  const e = new Error(`ENOENT: ${key}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

export async function read(
  accessor: DiscordAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const prefix = path.prefix
  let raw = path.original
  if (prefix !== '' && raw.startsWith(prefix)) {
    raw = raw.slice(prefix.length) || '/'
  }
  const key = stripSlash(raw)
  const parts = key.split('/')

  // <guild>/channels/<ch>/<date>/chat.jsonl
  if (
    parts.length === 5 &&
    parts[1] === 'channels' &&
    parts[4] === 'chat.jsonl' &&
    parts[0] !== undefined &&
    parts[2] !== undefined &&
    parts[3] !== undefined
  ) {
    if (index === undefined) throw fileNotFound(key)
    const chKey = `${parts[0]}/${parts[1]}/${parts[2]}`
    const chLookup = await index.get(`${prefix}/${chKey}`)
    if (chLookup.entry === undefined || chLookup.entry === null) throw fileNotFound(key)
    return await getHistoryJsonl(accessor, chLookup.entry.id, parts[3])
  }

  // <guild>/channels/<ch>/<date>/files/<blob>
  if (parts.length === 6 && parts[1] === 'channels' && parts[4] === 'files') {
    if (index === undefined) throw fileNotFound(key)
    const virtualKey = `${prefix}/${key}`
    let lookup = await index.get(virtualKey)
    if (lookup.entry === undefined || lookup.entry === null) {
      // Hydrate via date dir readdir (triggers fetchDay)
      const dateKey = parts.slice(0, 4).join('/')
      const dateVk = `${prefix}/${dateKey}`
      await discordReaddir(
        accessor,
        new PathSpec({ original: dateVk, directory: dateVk, prefix }),
        index,
      )
      lookup = await index.get(virtualKey)
    }
    if (lookup.entry === undefined || lookup.entry === null) throw fileNotFound(key)
    const extra = lookup.entry.extra
    const url =
      typeof extra.url === 'string' && extra.url !== ''
        ? extra.url
        : typeof extra.proxy_url === 'string'
          ? extra.proxy_url
          : ''
    if (url === '') throw fileNotFound(key)
    return await downloadFile(url)
  }

  // <guild>/members/<user>.json
  if (
    parts.length === 3 &&
    parts[1] === 'members' &&
    parts[2]?.endsWith('.json') === true &&
    parts[0] !== undefined
  ) {
    if (index === undefined) throw fileNotFound(key)
    const virtualKey = `${prefix}/${key}`
    const lookup = await index.get(virtualKey)
    if (lookup.entry === undefined || lookup.entry === null) throw fileNotFound(key)
    const guildLookup = await index.get(`${prefix}/${parts[0]}`)
    if (guildLookup.entry === undefined || guildLookup.entry === null) throw fileNotFound(key)
    const members = await listMembers(accessor, guildLookup.entry.id)
    for (const m of members) {
      if (m.user?.id === lookup.entry.id) {
        return encoder.encode(JSON.stringify(m))
      }
    }
    throw fileNotFound(key)
  }

  throw fileNotFound(key)
}

File: python/mirage/core/discord/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json

from mirage.accessor.discord import DiscordAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.discord.files import download_file
from mirage.core.discord.history import get_history_jsonl
from mirage.core.discord.members import list_members
from mirage.core.discord.readdir import readdir as _readdir
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def _ensure_channel(
    index: IndexCacheStore,
    prefix: str,
    ch_key: str,
    virtual: str,
):
    ch_virtual = prefix + "/" + ch_key
    lookup = await index.get(ch_virtual)
    if lookup.entry is None:
        raise enoent(virtual)
    return lookup


async def read(
    accessor: DiscordAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> bytes:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original if isinstance(path, PathSpec) else path
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.original

    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")
    parts = key.split("/")

    # <guild>/channels/<ch>/<date>/chat.jsonl
    if (len(parts) == 5 and parts[1] == "channels"
            and parts[4] == "chat.jsonl"):
        if index is None:
            raise enoent(virtual)
        ch_key = f"{parts[0]}/{parts[1]}/{parts[2]}"
        ch_lookup = await _ensure_channel(index, prefix, ch_key, virtual)
        return await get_history_jsonl(accessor.config, ch_lookup.entry.id,
                                       parts[3])

    # <guild>/channels/<ch>/<date>/files/<blob>
    if (len(parts) == 6 and parts[1] == "channels" and parts[4] == "files"):
        if index is None:
            raise enoent(virtual)
        virtual_key = prefix + "/" + key
        lookup = await index.get(virtual_key)
        if lookup.entry is None:
            # Hydrate via date dir readdir, which triggers _fetch_day
            date_key = "/".join(parts[:4])
            date_spec = PathSpec(
                original=prefix + "/" + date_key,
                directory=prefix + "/" + date_key,
                prefix=prefix,
            )
            await _readdir(accessor, date_spec, index)
            lookup = await index.get(virtual_key)
        if lookup.entry is None:
            raise enoent(virtual)
        url = (lookup.entry.extra
               or {}).get("url") or (lookup.entry.extra
                                     or {}).get("proxy_url") or ""
        if not url:
            raise enoent(virtual)
        return await download_file(url)

    # <guild>/members/<user>.json
    if len(parts) == 3 and parts[1] == "members":
        if index is None:
            raise enoent(virtual)
        virtual_key = prefix + "/" + key
        entry_lookup = await index.get(virtual_key)
        if entry_lookup.entry is None:
            raise enoent(virtual)
        guild_virtual = prefix + "/" + parts[0]
        guild_lookup = await index.get(guild_virtual)
        if guild_lookup.entry is None:
            raise enoent(virtual)
        members = await list_members(accessor.config, guild_lookup.entry.id)
        for m in members:
            user = m.get("user", {})
            if user.get("id") == entry_lookup.entry.id:
                return json.dumps(m, ensure_ascii=False,
                                  separators=(",", ":")).encode()
        raise enoent(virtual)

    raise enoent(virtual)
